compiler/goasm.go

```diff
@@ -0,0 +1,298 @@
+		movConstAsm = "mov ${0}, #%d"
+		callAsm = fmt.Sprintf("bl %#v", forwardName)
+		callConstraints = "={x0},{x0},~{x1},~{x2},~{x3},~{x4},~{x5},~{x6},~{x7},~{x8},~{x9},~{x10},~{x11},~{x12},~{x13},~{x14},~{x15},~{x16},~{x17},~{x19},~{x20},~{x21},~{x22},~{x23},~{x24},~{x25},~{x26},~{x27},~{x28},~{lr},~{q0},~{q1},~{q2},~{q3},~{q4},~{q5},~{q6},~{q7},~{q8},~{q9},~{q10},~{q11},~{q12},~{q13},~{q14},~{q15},~{q16},~{q17},~{q18},~{q19},~{q20},~{q21},~{q22},~{q23},~{q24},~{q25},~{q26},~{q27},~{q28},~{q29},~{q30},~{nzcv},~{ffr},~{vg},~{memory}"
+		if b.GOOS != "darwin" {
+			callConstraints += ",~{x18}"
+		}
+	default:
+		b.addError(b.fn.Pos(), "unknown architecture for Go assembly: "+b.archFamily())
+	}
+
+	// Initialize parameters, create entry block, etc.
+	b.createFunctionStart(true)
+
+	// Determine the stack layout that's used for the Go ABI.
+	// The layout roughly follows this convention (from low to high address):
+	//   - empty space (for the return pointer?)
+	//   - parameters
+	//   - return values
+	// More information can be found here (ABI0 is equivalent to the regabi
+	// without any integer or floating point registers):
+	// https://go.googlesource.com/go/+/refs/heads/master/src/cmd/compile/abi-internal.md
+	// We need to use size calculations as used by gc (the regular Go compiler)
+	// because that's what the assembly expects. It's usually the same as LLVM,
+	// but importantly differs on ARM where int64 is 32-bit aligned in gc but
+	// 64-bit aligned according to LLVM (and the ARM AAPCS).
+	sizes := types.SizesFor("gc", b.GOARCH)
+	var paramValues []llvm.Value
+	var paramFields []*types.Var
+	for _, param := range b.fn.Params {
+		value := b.getValue(param)
+		paramValues = append(paramValues, value)
+		paramFields = append(paramFields, types.NewField(token.NoPos, nil, param.Name(), param.Type(), false))
+	}
+	var resultFields []*types.Var
+	results := b.fn.Signature.Results()
+	for i := 0; i < results.Len(); i++ {
+		field := results.At(i)
+		resultFields = append(resultFields, types.NewField(token.NoPos, nil, "result_"+strconv.Itoa(i), field.Type(), false))
+	}
+	paramStruct := types.NewStruct(paramFields, nil)
+	stackStructFields := []*types.Var{
+		types.NewField(token.NoPos, nil, "offset", offsetField, false),
+		types.NewField(token.NoPos, nil, "params", paramStruct, false),
+		types.NewField(token.NoPos, nil, "align", types.NewArray(types.Typ[types.Uintptr], 0), false),
+		types.NewField(token.NoPos, nil, "results", types.NewStruct(resultFields, nil), false),
+	}
+	stackStruct := types.NewStruct(stackStructFields, nil)
+	stackStructOffsets := sizes.Offsetsof(stackStructFields)
+
+	// Create the alloca.
+	// WARNING: we're assuming here that this alloca will be the same as the
+	// stack pointer before the call in the assembly. This is not necessarily
+	// the case. But it seems to be working.
+	// I have tried many other approaches and this appears to be the most
+	// reliable way to do it.
```

Comment:
This is not great, and I'd like to experiment with other ways of doing this once this is in. My main priority has been compatibility: making sure Go assembly works at all.